### DISCUSSION TITLE
Cross compile using codespace from Linux to Windows

### DISCUSSION THREAD

**Discussion by redflitzi** [community/community#108328] [Codespaces] [answered]
### Select Topic Area

Question

### Body

I am currently developing a CLI tool on a codespace (using c# .NET 8). I am planning to compile it as Native AOT for both Linux and Windows, and perhaps also for Mac.

As cross-compiling AOT from Linux to Windows is currently not supported, I am looking for a alternative way to use msbuild on Windows with my current codebase. Because I work from changing locations I am using codespace VSCode in the browser, and avoid installing and maintaining .NET dev on each of the Windows desktops. Everything should ideally happen in the cloud, accessible for me from anywhere.

But I am lost, because I am such a noob in so many related things.
I even did not find the right place to ask.
Is it possible to somehow control Windows building process from .csproj inside the Linux codespace? Can I use a Windows Docker container only for the msbuild command? Or do I need Github Actions? I saw that there are minutes consumed, Windows needs twice as much as Linux, so is there a Windows container involved that can compile?
Or should I learn Azure? What else?
There are lots of detailed documentation to each of those themes, but I do not find how to put them together.

I feel that there must be many other developers that already solved this problem. But how?

So I would be very happy if someone pushes me in the right direction!

**Comment by spenserblack** [comment, marked as answer]
Unfortunately, [it doesn't seem like Codespaces will support Windows containers anytime soon](https://github.com/orgs/community/discussions/9977).

I'm not too familiar with C#, but to share what I've done, I develop in an Ubuntu codespace, and use actions to both run test and build release assets. See [this project's workflows](https://github.com/spenserblack/gengo/tree/c86bd8bd81c53791f21bd97aa0b833b8c4dcea99/.github/workflows) for some examples. As far as the minutes limit is concerned, you might want to build a Windows version only when creating a *release* asset, and do all your testing in Linux (assuming you don't have any intended OS-specific behavior). Note that you're also consuming storage and CPU time in your codespace, so using actions may be more efficient for testing and building.

You can definitely use docker to compile a Windows utility. See [devcontainer features](https://github.com/devcontainers/features/tree/main/src) to help with adding docker tools to your devcontainer. But, personally, I'd rather use an action than manually compile a project by using Docker.

**Reply by redflitzi** [reply]
Thank you very much, @spenserblack !

This helped me to find my way. I will use Github Actions with a windows-latest runner.
Via upload I'll get my published App (debug and release) back.
Just working on the workflows now.
Wont need Docker.

Thank you again.

**Reply by spenserblack** [reply]
No problem! Sounds like you might want to look into using the `actions/upload-artifact` action.

**Comment by redflitzi** [comment]
Thank you.
